### DISCUSSION TITLE
Is it possible to make ScoreboardInOrder more flexible

### DISCUSSION THREAD

**Discussion by calloftimereiver** [SpinalHDL/SpinalHDL#1540] [Ideas]
If I want to compare matrix data output in Stream form, that is, each fire will output the data of a matrix. I want to input the data type Array[Array[Int]] into ScoreboardInOrder and compare it, but due to the limitation of Array itself, I directly call 
 spinal.lib.sim.ScoreboardInOrder(lib/src/main/scala/spinal/lib/sim/Scoreboard.scala) with `def compare(ref : T, dut : T) = !(ref != dut)` function may not yield intuitive results, and I was wondering if the developers could consider adding overloading for this function to make it easier for users to customize the output impedance to achieve their desired results.

**Comment by Dolu1990** [comment]
Hi ^^ 

I pushed a fix in dev, allowing it to be overloaded.

**Reply by calloftimereiver** [reply]
Currently, I wrote like this:
```
case class ScoreboardInOrder_matrix() {
  val dut,ref = mutable.Queue[Array[Array[Int]]]()
  var matches = 0
 ...
  def compare(ref : Array[Array[Int]], dut : Array[Array[Int]]) = !(ref.deep != dut.deep)

  def printMatrix(matrix: Array[Array[Int]]): Unit = {
    matrix.foreach(row => println(row.mkString("\t")))
    println()
  }

  def check(): Unit ={
    if(ref.nonEmpty && dut.nonEmpty){
      val dutHead = dut.dequeue()
      val refHead = ref.dequeue()

      if(!compare(refHead, dutHead)){
        println("Transaction mismatch :")
        println("REF :")
        printMatrix(refHead)
        println("DUT :")
        printMatrix(dutHead)
        simFailure()
      }
      matches += 1
    }
  }
 ...
}
```
Can I just overload the original `ScoreboardInOrder[T]` like this, after the update?
```
class ScoreboardInOrder_matrix_new extends ScoreboardInOrder[Array[Array[Int]]] {
  override def compare(ref: Array[Array[Int]], dut: Array[Array[Int]]): Boolean = !(ref.deep != dut.deep)
  def printMatrix(matrix: Array[Array[Int]]): Unit = {
    matrix.foreach(row => println(row.mkString("\t")))
    println()
  }
  override def check(): Unit = {
    if (ref.nonEmpty && dut.nonEmpty) {
      val dutHead = dut.dequeue()
      val refHead = ref.dequeue()

      if (!compare(refHead, dutHead)) {
        println("Transaction mismatch :")
        println("REF :")
        printMatrix(refHead)
        println("DUT :")
        printMatrix(dutHead)
        simFailure()
      }
      matches += 1
    }
  }
}
```
I'm a beginner to scala, so please correct me if I do anything stupid

**Reply by Dolu1990** [reply]
Yes with that update in the dev branch you can overload things.
The only reason why you couldn't so far was that class was implemented as a "case class". That is lifted now ^^

**Reply by calloftimereiver** [reply]
OK, thank you.
